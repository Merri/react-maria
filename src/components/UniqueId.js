import React from 'react'
import PropTypes from 'prop-types'

export const uniqueIdSet = new Set()

export function clearUniqueIds() {
    uniqueIdSet.clear()
}

export function registerUniqueId(identifier) {
    let num = 1

    while (uniqueIdSet.has(identifier + num)) {
        num++
    }

    uniqueIdSet.add(identifier + num)
    return identifier + num
}

export function unregisterUniqueId(uniqueId) {
    if (uniqueIdSet.has(uniqueId)) {
        uniqueIdSet.delete(uniqueId)
    }
}

const html4Id = /^[A-Za-z]+[\w\-:.]*$/

export function withUniqueId(options) {
    const { identifier = 'maria-uid' } = options || {}

    // eslint-disable-next-line no-undef
    if (process.env.NODE_ENV !== 'production') {
        if (identifier == null || !html4Id.test(identifier)) {
            throw new Error('[withUniqueId] identifier must be a HTML4 compatible id string starting with ASCII letter')
        }
    }

    return function(Component) {
        class UniqueId extends React.PureComponent {
            static propTypes = {
                forwardedRef: PropTypes.any,
                uniqueId: PropTypes.string,
            }

            nextIds = {}
            state = {
                uniqueId: registerUniqueId(this.props.uniqueId != null ? this.props.uniqueId : identifier),
                uniqueIdProps: this.props.uniqueId,
            }

            static getDerivedStateFromProps(props, state) {
                if (state.uniqueIdProps === props.uniqueId) {
                    return null
                }

                unregisterUniqueId(state.uniqueId)

                return {
                    uniqueId: registerUniqueId(props.uniqueId != null ? props.uniqueId : identifier),
                    uniqueIdProps: props.uniqueId,
                }
            }

            componentWillUnmount() {
                this.uniqueIdGen()
                unregisterUniqueId(this.state.uniqueId)
            }

            lastUniqueId = id => {
                const key = typeof id === 'string' ? id : ''
                const array = this.nextIds[key]

                return (Array.isArray(array) && array[array.length - 1]) || null
            }

            nextUniqueId = id => {
                const key = typeof id === 'string' ? id : ''
                const uniqueId = registerUniqueId(`${this.state.uniqueId}.${key}`)

                if (!Array.isArray(this.nextIds[key])) {
                    this.nextIds[key] = []
                }

                this.nextIds[key].push(uniqueId)
                return uniqueId
            }

            generator = Object.defineProperty(Object.defineProperty({}, 'last', { value: this.lastUniqueId }), 'next', {
                value: this.nextUniqueId,
            })

            uniqueIdGen = () => {
                const keys = Object.keys(this.nextIds)

                for (let j = 0; j < keys.length; j++) {
                    const key = keys[j]

                    if (this.nextIds[key].length === 0) {
                        delete this.nextIds[key]
                    } else {
                        while (this.nextIds[key].length) {
                            unregisterUniqueId(this.nextIds[key].pop())
                        }
                    }
                }

                return this.generator
            }

            render() {
                const { forwardedRef, ...props } = this.props
                const { uniqueId } = this.state

                return <Component {...props} ref={forwardedRef} uniqueId={uniqueId} uniqueIdGen={this.uniqueIdGen} />
            }
        }
        // eslint-disable-next-line react/display-name,react/no-multi-comp
        return React.forwardRef((props, ref) => <UniqueId {...props} forwardedRef={ref} />)
    }
}
