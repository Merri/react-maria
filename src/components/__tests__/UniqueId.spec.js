/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { clearUniqueIds, registerUniqueId, uniqueIdSet, unregisterUniqueId, withUniqueId } from '../UniqueId'

function noop() {}

afterEach(() => {
    clearUniqueIds()
})

test('registerUniqueId adds new items to uniqueIdSet', () => {
    expect(registerUniqueId('test')).toEqual('test1')
    expect(registerUniqueId('test')).toEqual('test2')
    expect(uniqueIdSet.size).toEqual(2)
})

test('clearUniqueIds should clear uniqueIdSet', () => {
    registerUniqueId('test')
    expect(uniqueIdSet.size).toEqual(1)
    clearUniqueIds()
    expect(uniqueIdSet.size).toEqual(0)
})

test('unregisterUniqueId removes items from uniqueIdSet', () => {
    expect(registerUniqueId('test')).toEqual('test1')
    unregisterUniqueId('test1')
    expect(registerUniqueId('test')).toEqual('test1')
    unregisterUniqueId('test1')
    expect(uniqueIdSet.size).toEqual(0)
})

test('withUniqueId throws when identifier is not HTML4 specced id string', () => {
    global.spyOn(console, 'error')
    expect(() => withUniqueId({ identifier: '1' })).toThrowErrorMatchingSnapshot()
    expect(() => withUniqueId({ identifier: 1 })).toThrowErrorMatchingSnapshot()
    expect(() => withUniqueId({ identifier: null })).toThrowErrorMatchingSnapshot()
    expect(() => withUniqueId({ identifier: {} })).toThrowErrorMatchingSnapshot()
    expect(() => withUniqueId({ identifier: noop })).toThrowErrorMatchingSnapshot()
    expect(() => withUniqueId({ identifier: 'åäö' })).toThrowErrorMatchingSnapshot()
    expect(() => withUniqueId({ identifier: '$€' })).toThrowErrorMatchingSnapshot()
    expect(() => withUniqueId({ identifier: '[(' })).toThrowErrorMatchingSnapshot()
})

test('withUniqueId acts as a HoC for components', () => {
    const data = {}

    const Component = withUniqueId()(
        React.forwardRef(props => {
            data.props = props
            return <div />
        }),
    )

    const tree = renderer.create(<Component />)
    expect(data.props.uniqueId).toEqual('maria-uid1')
    expect(data.props.uniqueIdGen).toBeNull()
    tree.update(<Component uniqueId="test" />)
    expect(data.props.uniqueId).toEqual('test1')
    tree.update(<Component />)
    expect(data.props.uniqueId).toEqual('maria-uid1')
})

test('withUniqueId exposes ID generator for components when requested', () => {
    const Component = withUniqueId({ includeGenerator: true })(
        React.forwardRef(props => {
            const uid = props.uniqueIdGen()

            return (
                <ul>
                    {['a', 'b'].map((subId, index) => (
                        <li key={index} id={uid.next(props.useSubId && subId)}>
                            {`I am known as "${uid.last(props.useSubId && subId)}"`}
                        </li>
                    ))}
                </ul>
            )
        }),
    )

    const tree = renderer.create(<Component />)
    expect(tree.root.findByProps({ id: 'maria-uid1.1' }).children).toEqual(['I am known as "maria-uid1.1"'])
    expect(tree.root.findByProps({ id: 'maria-uid1.2' }).children).toEqual(['I am known as "maria-uid1.2"'])
    expect(uniqueIdSet.size).toEqual(3)
    tree.update(<Component useSubId />)
    expect(tree.root.findByProps({ id: 'maria-uid1.a1' }).children).toEqual(['I am known as "maria-uid1.a1"'])
    expect(tree.root.findByProps({ id: 'maria-uid1.b1' }).children).toEqual(['I am known as "maria-uid1.b1"'])
    expect(uniqueIdSet.size).toEqual(3)
})
