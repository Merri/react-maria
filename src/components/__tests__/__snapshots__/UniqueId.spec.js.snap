// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withUniqueId exposes ID generator for components when requested and does not get confused 1`] = `
<dl
  id="maria-uid1"
>
  <dt
    id="maria-uid1.1"
  >
    I am known as "maria-uid1.1"
  </dt>
  <dd>
    I am here to make semantics right.
  </dd>
  <dt
    id="maria-uid1.2"
  >
    I am known as "maria-uid1.2"
  </dt>
  <dd>
    I am here to make semantics right.
  </dd>
  <dd />
</dl>
`;

exports[`withUniqueId exposes ID generator for components when requested and does not get confused 2`] = `
<dl
  id="maria-uid1"
>
  <dt
    id="maria-uid1.1"
  >
    I am known as "maria-uid1.1"
  </dt>
  <dd>
    I am here to make semantics right.
  </dd>
  <dd
    id="maria-uid1.a1"
  >
    maria-uid1.a1
  </dd>
  <dd
    id="maria-uid1.a2"
  >
    maria-uid1.a2
  </dd>
  <dt
    id="maria-uid1.2"
  >
    I am known as "maria-uid1.2"
  </dt>
  <dd>
    I am here to make semantics right.
  </dd>
  <dd
    id="maria-uid1.b1"
  >
    maria-uid1.b1
  </dd>
  <dd
    id="maria-uid1.b2"
  >
    maria-uid1.b2
  </dd>
  <dd />
</dl>
`;

exports[`withUniqueId exposes ID generator for components when requested and does not get confused 3`] = `
<dl
  id="maria-uid1"
>
  <dt
    id="maria-uid1.1"
  >
    I am known as "maria-uid1.1"
  </dt>
  <dd>
    I am here to make semantics right.
  </dd>
  <dd
    id="maria-uid1.a1"
  >
    maria-uid1.a1
  </dd>
  <dd
    id="maria-uid1.a2"
  >
    maria-uid1.a2
  </dd>
  <dt
    id="maria-uid1.2"
  >
    I am known as "maria-uid1.2"
  </dt>
  <dd>
    I am here to make semantics right.
  </dd>
  <dd
    id="maria-uid1.b1"
  >
    maria-uid1.b1
  </dd>
  <dd
    id="maria-uid1.b2"
  >
    maria-uid1.b2
  </dd>
  <dd>
    <dl
      id="test1"
    >
      <dt
        id="test1.1"
      >
        I am known as "test1.1"
      </dt>
      <dd>
        I am here to make semantics right.
      </dd>
      <dd
        id="test1.a1"
      >
        test1.a1
      </dd>
      <dd
        id="test1.a2"
      >
        test1.a2
      </dd>
      <dt
        id="test1.2"
      >
        I am known as "test1.2"
      </dt>
      <dd>
        I am here to make semantics right.
      </dd>
      <dd
        id="test1.b1"
      >
        test1.b1
      </dd>
      <dd
        id="test1.b2"
      >
        test1.b2
      </dd>
      <dd />
    </dl>
  </dd>
</dl>
`;

exports[`withUniqueId exposes ID generator for components when requested and does not get confused 4`] = `
<dl
  id="maria-uid2"
>
  <dt
    id="maria-uid2.1"
  >
    I am known as "maria-uid2.1"
  </dt>
  <dd>
    I am here to make semantics right.
  </dd>
  <dt
    id="maria-uid2.2"
  >
    I am known as "maria-uid2.2"
  </dt>
  <dd>
    I am here to make semantics right.
  </dd>
  <dd />
</dl>
`;

exports[`withUniqueId throws when identifier is not HTML4 specced id string 1`] = `"[withUniqueId] identifier must be a HTML4 compatible id string starting with ASCII letter"`;

exports[`withUniqueId throws when identifier is not HTML4 specced id string 2`] = `"[withUniqueId] identifier must be a HTML4 compatible id string starting with ASCII letter"`;

exports[`withUniqueId throws when identifier is not HTML4 specced id string 3`] = `"[withUniqueId] identifier must be a HTML4 compatible id string starting with ASCII letter"`;

exports[`withUniqueId throws when identifier is not HTML4 specced id string 4`] = `"[withUniqueId] identifier must be a HTML4 compatible id string starting with ASCII letter"`;

exports[`withUniqueId throws when identifier is not HTML4 specced id string 5`] = `"[withUniqueId] identifier must be a HTML4 compatible id string starting with ASCII letter"`;

exports[`withUniqueId throws when identifier is not HTML4 specced id string 6`] = `"[withUniqueId] identifier must be a HTML4 compatible id string starting with ASCII letter"`;

exports[`withUniqueId throws when identifier is not HTML4 specced id string 7`] = `"[withUniqueId] identifier must be a HTML4 compatible id string starting with ASCII letter"`;

exports[`withUniqueId throws when identifier is not HTML4 specced id string 8`] = `"[withUniqueId] identifier must be a HTML4 compatible id string starting with ASCII letter"`;
